async_retry.go

```diff
@@ -0,0 +1,134 @@
+package asyncretry
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/avast/retry-go/v4"
+)
+
+type AsyncRetryFunc func(ctx context.Context) error
+
+type AsyncRetry interface {
+	// Do calls f and retry if necessary.
+	// In most cases, you should call Do in a new goroutine.
+	Do(ctx context.Context, f AsyncRetryFunc, opts ...Option) error
+	// Shutdown shutdowns gracefully
+	Shutdown(ctx context.Context) error
+}
+
+type asyncRetry struct {
+	// FIXME use RWMutex instead
+	mu           sync.Mutex // guards wg and shutdownChan
+	wg           sync.WaitGroup
+	shutdownChan chan struct{}
+}
+
+func NewAsyncRetry() AsyncRetry {
+	return &asyncRetry{
+		wg:           sync.WaitGroup{},
+		shutdownChan: make(chan struct{}),
+	}
+}
+
+var InShutdownErr = fmt.Errorf("AsyncRetry is in shutdown")
+
+func (a *asyncRetry) Do(ctx context.Context, f AsyncRetryFunc, opts ...Option) (retErr error) {
+	a.mu.Lock()
+	select {
+	case <-a.shutdownChan:
+		return InShutdownErr
```

Comment:
It seems that mutex is not Unlocked when shutdownChan is already closed.
This means that if a `asyncRetry.Do` is called twice on `asyncRetry resource that has already been shutdown, the second call will fail to acquire a `mu.Lock` and will result in a deadlock.

The problem only occurs if the library user continues to mistakenly use a resource that has been shutdown, but it seems like a potential deadlock that should be fixed.

Reply:
Thx!
You are right!
I've just submitted for this bug: https://github.com/Kyash/async-retry/pull/4